WcaOnRails/app/helpers/competitions_helper.rb

```diff
@@ -2,26 +2,39 @@
 
 module CompetitionsHelper
   def competition_message_for_user(competition, user)
-    messages = []
-    registration = competition.registrations.find_by_user_id(user.id)
-    if competition.cancelled?
-      messages << t('competitions.messages.cancelled')
-    elsif registration
-      # Display a tooltip on a user's bookmarked competition based on their registration status.
-      if registration.accepted?
-        messages << t('competitions.messages.tooltip_registered')
-      elsif registration.deleted?
-        messages << t('competitions.messages.tooltip_deleted')
-      else # If not delted or accepted, assume user is on the waiting list
-        messages << t('competitions.messages.tooltip_waiting_list')
-      end
+    # Generates a list of messages, which will be combined and displayed in a tooltip to the user in their bookmarked
+    # competitions list when they hover over a competition.
+    # Message indicates the state of the competition, and the state of the user's registration.
+
+    messages = [] # Messages to be combined
+
+    if competition.cancelled? # If the competition is cancelled, just return the cancellation message to the user.
+      return t('competitions.messages.cancelled')
+    end
+
+    # Registration state messages
```

Comment:
A good rule of thumb is that if you need a comment to describe the next bunch of code, it's often a good idea to extract that code into a helper function and give it a descriptive name. If the name is well-chosen, it eliminates the need for the comment - and makes the original function easier to understand at a glance because it's shorter and calls a well-named helper function.

I'm not saying it's necessary to make such changes in this case, but if the rest of the function below here just called a helper `addRegistrationStatusMessageIfRegistered` then called another helper `addCompetitionVisibilityMessage` then returned `messages.join()`, it would be super easy to understand.

Although I'm not very familiar with Ruby, so someone correct me if that's an un-Ruby way of doing things.

Reply:
Very interesting idea, thanks! I did what you suggested in terms of helper functions, mostly as a learning exercise. Would be curious on your feedback! And very happy if your feedback is "this is worse, revert it" - I'm not convinced of it myself.

If I was in Python I would have defined the helper functions as inline functions, to minimise clutter in the rest of the file. Ruby doesn't allow this, so I've defined them as separate functions. Curious what your thoughts are on this - I think it makes the `competition_message_for_user` function much easier to understand, but makes `competitions_helper.rb` as a whole look more cluttered? Perhaps we could define these as private methods to help alleviate that?

Reply:
I like it this way, with the helper functions. (I'm not familiar with Python.) Only thing I would change is putting the primary function, `get_competition_status_message`, above the helpers. So the most important thing is at the top, and if you want more details you can continue reading down. (Unless Ruby would complain about using something before it's defined, in which case ignore this.) If these helpers aren't intended to be used outside then making them private makes sense to me.

Reply:
Thanks! Functions need to be defined before they're called in Ruby, and I agree it looks gross. Luckily `private` lets us kill two birds with one stone here - we can make stuff private _and_ get them out the way: 
* I've moved the helper functions into a `private` section at the bottom of the file, which is the Ruby convention anyway.
* Also moved some helpers elsewhere in the file with inline `private` definitions, which isn't great Ruby convention

Aside: In Python, I could do the following (I think, I haven't played with nested function definitions much):

```python
def main_method:
  messages = []
  messages.append(helper_function_1())
  messages.append(helper_function_2())

  def helper_function_1():
    # Do some stuff
    return

  def helper_function_2():
    # Do some stuff
    return

  return messages